eos-boot-upgrade

```diff
@@ -0,0 +1,278 @@
+#!/usr/bin/python3 -u
+#
+# Copyright (C) 2018 Endless Mobile, Inc.
+# Author:
+#  Carlo Caione <carlo@endlessm.com>
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
+# USA.
+
+import contextlib
+import distutils.dir_util
+import filecmp
+import json
+import os
+import re
+import shutil
+import subprocess
+
+
+class BootUpgradeException(Exception):
+    def __init__(self, *args):
+        super(BootUpgradeException, self).__init__(args)
+        self.msg = ' '.join(map(str, args))
+
+    def __str__(self):
+        return str(self.msg)
+
+
+class Partition(object):
+    def __init__(self, part_number, disk, **kwargs):
+        self.disk = disk
+        self.number = part_number
+
+        for config_name in kwargs:
+            setattr(self, config_name, kwargs[config_name])
+
+    @staticmethod
+    def parse(config, disk):
+        '''Parse the partition data passed by sfdisk json data'''
+        if 'node' not in config:
+            raise BootUpgradeException('"node" missing from sfdisk partition info:',
+                                       config)
+
+        part_number = re.match('^.*([0-9]+)$',
+                               config['node']).group(1)
+        part_config = {
+            'device': config['node'],
+            'uuid': config.get('uuid'),
+            'type': config.get('type')
+        }
+        return Partition(part_number, disk, **part_config)
+
+
+class Disk(object):
+    def __init__(self, device):
+        self.label = None
+        self.partitions = {}
+        self.device = device
+        self._read_partition_table()
+
+    def _read_partition_table(self):
+        '''Read partition table using sfdisk json data'''
+        sfdisk_json = subprocess.check_output(['sfdisk',
+                                               '--json',
+                                               self.device]).decode('utf-8')
+        disk_config = json.loads(sfdisk_json)
+
+        self.label = disk_config['partitiontable']['label'] or None
+
+        for part_config in disk_config['partitiontable']['partitions']:
+            part = Partition.parse(part_config, self)
+            self.partitions[part.number] = part
+
+    def get_partition_by_guid(self, guid):
+        '''Get partition by GUID / partition type'''
+        for part_num, conf in self.partitions.items():
```

Comment:
part_num is unused – you can do this instead:

```python
for conf in self.partitions.values():
    ...
```

Reply:
In fact, since we never index `self.partitions` by its actual key (the partition number), `self.partitions` could just be a list rather than a dictionary.